subiquitycore/model.py

```diff
@@ -0,0 +1,24 @@
+# Copyright 2021 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from abc import ABC
+import logging
+
+log = logging.getLogger("subiquitycore.model")
+
+
+class CloudConfigModel(ABC):
+    def make_cloudconfig(self):
```

Comment:
So I'm not sure this adds anything. There are a few thoughts it makes bounce around in my head:

 1. if you have a base class like this, then the check in subiquity.models.subiquity should be "isinstance" not "getattr"
 2. there's not much point inheriting from ABC if you don't have an abstractmethod 
 3. as you've probably noticed while doing this, not all postinstall models contribute to cloud-init config and some are not custom types at all.
 4. this should be in subiquity not subiquitycore

all that said, I think I would support just dropping this baseclass and continuing to to the getattr based test in subiquity.models.subiquity, but I'm also not opposed to having this baseclass so long as it's in the subiquity package, has an abstractmethod and you use isinstance in subiquity.models.subiquity

Reply:
An earlier version of this PR removed "packages" and "userdata" from POSTINSTALL_MODEL_NAMES, and thus could skip the getattr(model, 'make_cloudconfig') check.  Does POSTINSTALL_MODEL_NAMES need those two entries?

Reply:
in principle yes, because we don't want to move on to the postinstall stages until they have been configured. We could divide up the postinstall steps to be more granular -- so only the coroutine that's installing the packages blocks on packages being configured etc but ehhhh. Basically "postinstall" is wider than just cloud-init config generation whereas "install" really is just making the curtin config, I guess.

In practice I'm sure everything works if they are not in POSTINSTALL_MODEL_NAMES because both are always noninteractive and so get configured very early.